vm/actor/src/builtin/miner/mod.rs

```diff
@@ -94,90 +121,128 @@ impl Actor {
+
+        let empty_deadline_cid = rt
+            .store()
+            .put(&Deadline::new(empty_array.clone()), Blake2b256)
+            .unwrap();
+
+        let empty_deadlines_cid = rt
+            .store()
+            .put(&Deadlines::new(empty_deadline_cid), Blake2b256)
+            .unwrap();
+
+        let empty_vesting_funds_cid = rt.store().put(&VestingFunds::new(), Blake2b256).unwrap();
 
         let current_epoch = rt.curr_epoch();
         let blake2b = |b: &[u8]| rt.syscalls().hash_blake2b(b);
         let offset = assign_proving_period_offset(*rt.message().receiver(), current_epoch, blake2b)
             .map_err(|e| {
-                ActorError::new(
-                    ExitCode::ErrSerialization,
-                    format!("failed to assign proving period offset: {}", e),
+                actor_error!(
+                    ErrSerialization,
+                    "failed to assign proving period offset: {}",
+                    e
                 )
             })?;
 
         let period_start = next_proving_period_start(current_epoch, offset);
         assert!(period_start > current_epoch);
 
-        let st = State::new(
-            empty_array,
-            empty_map,
-            empty_deadlines_cid,
+        let info = MinerInfo::new(
             owner,
             worker,
+            control_addresses,
             params.peer_id,
             params.multi_address,
             params.seal_proof_type,
-            period_start,
         )
-        .map_err(|e| ActorError::new(ExitCode::ErrIllegalArgument, e))?;
+        .map_err(|e| {
+            actor_error!(
+                ErrIllegalArgument,
+                "failed to construct initial miner info: {}",
+                e
+            )
+        })?;
+        let info_cid = rt.store().put(&info, Blake2b256).unwrap();
+
+        let st = State::new(
+            info_cid,
+            period_start,
+            empty_bitfield_cid,
+            empty_array,
+            empty_map,
+            empty_deadlines_cid,
+            empty_vesting_funds_cid,
+        );
         rt.create(&st)?;
 
-        // Register cron callback for epoch before the first proving period starts.
+        // Register first cron callback for epoch before the first proving period starts.
         enroll_cron_event(
             rt,
             period_start - 1,
             CronEventPayload {
-                event_type: CRON_EVENT_PROVING_PERIOD,
-                sectors: Default::default(),
+                event_type: CRON_EVENT_PROVING_DEADLINE,
             },
         )?;
 
         Ok(())
     }
 
+    fn get_miner_info<BS, RT>(rt: &mut RT, state: &State) -> Result<MinerInfo, ActorError>
+    where
+        BS: BlockStore,
+        RT: Runtime<BS>,
+    {
+        state
+            .get_info(rt.store())
+            .map_err(|e| actor_error!(ErrIllegalState, "could not read miner info: {}", e))
```

Comment:
This should be a downcast, not a mapping (The exit code passed into `RequireNoErr` on go impl is only a default